### PR TITLE
add multiple resources

### PR DESCRIPTION
**Pull request by dbasunag**

### CODE REVIEW

**Comment by dbasunag** [comment]
!verified

File: ocp_resources/cron_job.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        self.schedule = schedule
+
+    def to_dict(self):
+        super().to_dict()
+        if not self.yaml_file:
+            self.res.update(
+                {
+                    "spec": {
+                        "jobTemplate": self.job_template,
+                        "schedule": self.schedule,
```
Comment (rnetser) [inline ocp_resources/cron_job.py:58]: can they be None? if not, need to verify values were provided

File: ocp_resources/cron_job.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+from ocp_resources.constants import TIMEOUT_4MINUTES
+from ocp_resources.resource import NamespacedResource
+
+
+class CronJob(NamespacedResource):
+    """
+    CronJob object. API reference:
+    https://docs.openshift.com/container-platform/4.12/rest_api/workloads_apis/cronjob-batch-v1.html
```
Comment (rnetser) [inline ocp_resources/cron_job.py:8]: please add a link to the api doc (where all fields are described)

File: ocp_resources/csi_storage_capacity.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+from ocp_resources.constants import TIMEOUT_4MINUTES
+from ocp_resources.resource import NamespacedResource
+
+
+class CSIStorageCapacity(NamespacedResource):
+    """
+    CSIStorageCapacity object. API reference:
+    https://docs.openshift.com/container-platform/4.12/rest_api/storage_apis/storage-apis-index.html#csistoragecapacity-storage-k8s-iov1
```
Comment (rnetser) [inline ocp_resources/csi_storage_capacity.py:8]: ditto (https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/csi-storage-capacity-v1/#CSIStorageCapacity)

File: ocp_resources/csi_storage_capacity.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    ):
+        """
+        Args:
+            name (str): Name of the endpoints resource
+            namespace (str): Namespace of endpoints resource
+            client: (DynamicClient): DynamicClient for api calls
+            capacity (str): value reported by the csi driver
+            maximum_volume_size (str): maximum volume size reported by csi driver
+            storage_class_name (str): storage class name
+            node_topology (object): defines which node has access to the storage for which capacity was reported
```
Comment (rnetser) [inline ocp_resources/csi_storage_capacity.py:36]: the type is `LabelSelector`, can you add an example?

File: ocp_resources/cron_job.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        Args:
+            name (str): Name of the endpoints resource
+            namespace (str): Namespace of endpoints resource
+            client: (DynamicClient): DynamicClient for api calls
+            teardown (bool): Indicates if the resource should be torn down at the end
+            privileged_client (DynamicClient): Privileged client for api calls
+            yaml_file (str): yaml file for the resource.
+            delete_timeout (int): timeout associated with delete action
+            schedule (str): schedule of the cron job
+            jobTemplate (object): describes data a job should have when created from a template
```
Comment (rnetser) [inline ocp_resources/cron_job.py:36]:
```suggestion
            job_template (object): describes data a job should have when created from a template
```

Comment (rnetser) [reply ocp_resources/cron_job.py:36]: can you add an example?

File: ocp_resources/csi_storage_capacity.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    def to_dict(self):
+        super().to_dict()
+        if not self.yaml_file:
+            self.res.update(
+                {
+                    "maximumVolumeSize": self.maximum_volume_size,
+                    "nodeTopology": self.node_topology,
+                    "storageClassName": self.storage_class_name,
+                    "capacity": self.capacity,
+                }
```
Comment (rnetser) [inline ocp_resources/csi_storage_capacity.py:66]: can be None?

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/csi_storage_capacity.py (hunk 3 above)
Comment (dbasunag) [reply ocp_resources/csi_storage_capacity.py:8]: Done and thank you!

File: ocp_resources/csi_storage_capacity.py (hunk 4 above)
Comment (dbasunag) [reply ocp_resources/csi_storage_capacity.py:36]: I will add example.

File: ocp_resources/cron_job.py (hunk 1 above)
Comment (dbasunag) [reply ocp_resources/cron_job.py:58]: No these are required params. I need to validate the values are provided and add the optional ones.

File: ocp_resources/csi_storage_capacity.py (hunk 6 above)
Comment (dbasunag) [reply ocp_resources/csi_storage_capacity.py:66]: No a storage class name is a required field. I will add necessary check for all the resources.

File: ocp_resources/cron_job.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        self.concurrency_policy = concurrency_policy
+        self.suspend = suspend
+        self.successful_jobs_history_limit = successful_jobs_history_limit
+        self.failed_jobs_history_limit = failed_jobs_history_limit
+        self.starting_deadline_seconds = starting_deadline_seconds
+
+    def to_dict(self):
+        super().to_dict()
+        if not self.yaml_file:
+            if not (self.job_template and self.schedule):
```
Comment (rnetser) [inline ocp_resources/cron_job.py:76]: what about the other ones? (concurrency_policy,failed_jobs_history_limit, etc)
either add to validation or mark as optional in the docstring

File: ocp_resources/csi_storage_capacity.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            maximum_volume_size (str): maximum volume size reported by csi driver
+            storage_class_name (str): storage class name
+            node_topology (LabelSelector): defines which node has access to the storage for which capacity was reported
+            teardown (bool): Indicates if the resource should be torn down at the end
+            privileged_client (DynamicClient): Privileged client for api calls
+            yaml_file (str): yaml file for the resource.
+            delete_timeout (int): timeout associated with delete action
+
+        Example:
+            node_topology: {'matchLabels': {'topology.hostpath.csi/node': 'c01-dbn-413-4c48b-worker-0-pmtv8'}}
```
Comment (rnetser) [inline ocp_resources/csi_storage_capacity.py:43]: please move under `node_topology` arg

File: ocp_resources/csi_storage_capacity.py (hunk 6 above)
Comment (rnetser) [reply ocp_resources/csi_storage_capacity.py:73]: what about the other ones? (either check or add `optional` to docsting)

File: ocp_resources/endpoint_slice.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            **kwargs,
+        )
+        self.address_type = address_type
+        self.endpoints = endpoints
+        self.ports = ports
+
+    def to_dict(self):
+        super().to_dict()
+        if not self.yaml_file:
+            if not (self.address_type and self.endpoints):
```
Comment (rnetser) [inline ocp_resources/endpoint_slice.py:57]: ports as well?

File: ocp_resources/lease.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        )
+        self.holder_identity = holder_identity
+        self.lease_transitions = lease_transitions
+        self.lease_duration_seconds = lease_duration_seconds
+        self.acquire_time = acquire_time
+        self.renew_time = renew_time
+
+    def to_dict(self):
+        super().to_dict()
+        if not self.yaml_file:
```
Comment (rnetser) [inline ocp_resources/lease.py:62]: same here - required params or optional

File: ocp_resources/pod_metrics.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+            delete_timeout=delete_timeout,
+            **kwargs,
+        )
+        self.containers = containers
+        self.timestamp = timestamp
+        self.window = window
+
+    def to_dict(self):
+        super().to_dict()
+        if not self.yaml_file:
```
Comment (rnetser) [inline ocp_resources/pod_metrics.py:57]: ditto

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/csi_storage_capacity.py (hunk 6 above)
Comment (dbasunag) [reply ocp_resources/csi_storage_capacity.py:73]: I have added the check for only the required ones. But I think adding `optional` to the docstring is a good idea. I will do this for all the resources I am adding. Thank you!

File: ocp_resources/lease.py (hunk 10 above)
Comment (dbasunag) [reply ocp_resources/lease.py:62]: None of the params are marked as required, based on https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/lease-v1/ - which to be honest is a bit confusing.

File: ocp_resources/pod_metrics.py (hunk 11 above)
Comment (dbasunag) [reply ocp_resources/pod_metrics.py:57]: Same here. I have not been able to find the actual api doc for PodMetrics. the closest information that I found was https://docs.openshift.com/container-platform/4.12/nodes/pods/nodes-pods-autoscaling.html - so far none of the params seems to be required.

File: ocp_resources/cron_job.py (hunk 7 above)
Comment (dbasunag) [reply ocp_resources/cron_job.py:76]: None of those are required. Only job_template and schedule are required. But I would need to update docstring to reflect that.

File: ocp_resources/pod_metrics.py (hunk 11 above)
Comment (dbasunag) [reply ocp_resources/pod_metrics.py]: Removing all these based on Ruth's recommendations.

**Comment by dbasunag** [comment]
!verified

File: ocp_resources/cron_job.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+            self.res.update(
+                {
+                    "spec": {
+                        "jobTemplate": self.job_template,
+                        "schedule": self.schedule,
+                        "concurrencyPolicy": self.concurrency_policy,
+                        "failedJobsHistoryLimit": self.failed_jobs_history_limit,
+                        "successfulJobsHistoryLimit": self.successful_jobs_history_limit,
+                        "suspend": self.suspend,
+                        "timezone": self.timezone,
```
Comment (rnetser) [inline ocp_resources/cron_job.py:90]: please add only if passed; i.e
```
if self.xxx: ....
```

File: ocp_resources/csi_storage_capacity.py (hunk 6 above)
Comment (rnetser) [reply ocp_resources/csi_storage_capacity.py:73]: please also add to spec if they are actually passed (see my comment above)

File: ocp_resources/endpoint_slice.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+                raise ValueError(
+                    "yaml_file or parameters 'address_type' and 'endpoints' are required."
+                )
+            self.res.update(
+                {
+                    "addressTypes": self.address_type,
+                    "endpoints": self.endpoints,
+                    "ports": self.ports,
```
Comment (rnetser) [inline ocp_resources/endpoint_slice.py:65]: add only if passed

File: ocp_resources/lease.py (hunk 10 above)
Comment (rnetser) [reply ocp_resources/lease.py:62]: also, please add to `spec` only those that are actually passed

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/cron_job.py (hunk 12 above)
Comment (dbasunag) [reply ocp_resources/cron_job.py:90]: Will do this for all resources I am adding (except pod_metrics).

File: ocp_resources/csi_storage_capacity.py (hunk 6 above)
Comment (dbasunag) [reply ocp_resources/csi_storage_capacity.py:73]: No spec for CSIStorageCapacity, but I understand what you meant. I will implement it.

File: ocp_resources/cron_job.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+            self.res.update(
+                {
+                    "spec": {
+                        "jobTemplate": self.job_template,
+                        "schedule": self.schedule,
+                    }
+                }
+            )
+            if self.timezone:
+                self.res["spec"]["timezone"] = self.timezone
```
Comment (rnetser) [inline ocp_resources/cron_job.py:90]: `self.res["spec"]` - please save and re-use

File: ocp_resources/lease.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+            if self.acquire_time:
+                self.res["spec"]["acquireTime"] = self.acquire_time
```
Comment (rnetser) [inline ocp_resources/lease.py:64]: ditto

**Review by rnetser — CHANGES_REQUESTED** [review]

**Review by rnetser — DISMISSED** [review]

File: ocp_resources/cron_job.py [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+        Args:
+            name (str): Name of the endpoints resource
+            namespace (str): Namespace of endpoints resource
+            client: (DynamicClient): DynamicClient for api calls
+            teardown (bool): Indicates if the resource should be torn down at the end
+            privileged_client (DynamicClient): Privileged client for api calls
+            yaml_file (str): yaml file for the resource.
+            delete_timeout (int, optional): timeout associated with delete action
+            schedule (str): schedule of the cron job
+            job_template (JobTemplateSpec): describes the job that would be created when a cronjob would be executed
```
Comment (myakove) [inline ocp_resources/cron_job.py:42]: `job_template` expect to get `JobTemplateSpec` object?

File: ocp_resources/csi_storage_capacity.py [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    ):
+        """
+        Args:
+            name (str): Name of the endpoints resource
+            namespace (str): Namespace of endpoints resource
+            client: (DynamicClient): DynamicClient for api calls
+            capacity (str, optional): value reported by the csi driver
+            maximum_volume_size (str, optional): maximum volume size reported by csi driver
+            storage_class_name (str): storage class name
+            node_topology (LabelSelector, optional): defines which node has access to the storage for which capacity
```
Comment (myakove) [inline ocp_resources/csi_storage_capacity.py:36]: `JobTemplateSpec` get `LabelSelector` object or dict?

File: ocp_resources/cron_job.py [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+        successful_jobs_history_limit=None,
+        failed_jobs_history_limit=None,
+        privileged_client=None,
+        yaml_file=None,
+        delete_timeout=TIMEOUT_4MINUTES,
+        **kwargs,
+    ):
+        """
+        Args:
+            name (str): Name of the endpoints resource
```
Comment (myakove) [inline ocp_resources/cron_job.py:34]: This is not `endpoints resource`, please fix the docstring.

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/cron_job.py (hunk 16 above)
Comment (dbasunag) [reply ocp_resources/cron_job.py:42]: As per the API doc, it is JobTemplateSpec type. https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/cron-job-v1/#CronJobSpec. I see both metadata and spec present,

File: ocp_resources/csi_storage_capacity.py (hunk 17 above)
Comment (dbasunag) [reply ocp_resources/csi_storage_capacity.py:36]: You mean node_topology? It is a dict.

**Review by rnetser — DISMISSED** [review]

File: ocp_resources/cron_job.py (hunk 16 above)
Comment (myakove) [reply ocp_resources/cron_job.py:42]: the docstring specify `JobTemplateSpec`, meaning you expect the user to sent `JobTemplateSpec` object here.

File: ocp_resources/csi_storage_capacity.py (hunk 17 above)
Comment (myakove) [reply ocp_resources/csi_storage_capacity.py:36]: Then it should be `node_topology (dict, optional)`
please fix where needed

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/cron_job.py (hunk 16 above)
Comment (dbasunag) [reply ocp_resources/cron_job.py:42]: I will change this to dict. It should be a dict

File: ocp_resources/csi_storage_capacity.py (hunk 17 above)
Comment (dbasunag) [reply ocp_resources/csi_storage_capacity.py:36]: I will update.

**Comment by dbasunag** [comment]
!verified

**Review by rnetser — DISMISSED** [review]

File: ocp_resources/pod_metrics.py [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+
+
+class PodMetrics(NamespacedResource):
+    """
+    PodMetrics object. API reference:
+    This resource is managed by: https://github.com/kubernetes-sigs/metrics-server and is a built-in resource
+    """
+
+    api_group = NamespacedResource.ApiGroup.METRICS_K8S_IO
+    api_version = NamespacedResource.ApiVersion.V1BETA1
```
Comment (myakove) [inline ocp_resources/pod_metrics.py:11]: api_version is not needed, we get it from the cluster.

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/pod_metrics.py (hunk 19 above)
Comment (dbasunag) [reply ocp_resources/pod_metrics.py]: Done.

**Review by rnetser — APPROVED** [review]

**Review by myakove — APPROVED** [review]

**Review by myakove — APPROVED** [review]

**Pull request merged**